sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala

```diff
@@ -86,28 +86,11 @@ case class ArrowEvalPythonExec(udfs: Seq[PythonUDF], resultAttrs: Seq[Attribute]
       sessionLocalTimeZone,
       pythonRunnerConf).compute(batchIter, context.partitionId(), context)
 
-    new Iterator[InternalRow] {
-
-      private var currentIter = if (columnarBatchIter.hasNext) {
-        val batch = columnarBatchIter.next()
-        val actualDataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType())
-        assert(outputTypes == actualDataTypes, "Invalid schema from pandas_udf: " +
-          s"expected ${outputTypes.mkString(", ")}, got ${actualDataTypes.mkString(", ")}")
-        batch.rowIterator.asScala
-      } else {
-        Iterator.empty
-      }
-
-      override def hasNext: Boolean = currentIter.hasNext || {
-        if (columnarBatchIter.hasNext) {
-          currentIter = columnarBatchIter.next().rowIterator.asScala
-          hasNext
-        } else {
-          false
-        }
-      }
-
-      override def next(): InternalRow = currentIter.next()
+    columnarBatchIter.flatMap { batch =>
+      val actualDataTypes = (0 until batch.numCols()).map(i => batch.column(i).dataType())
+      assert(outputTypes == actualDataTypes, "Invalid schema from pandas_udf: " +
+        s"expected ${outputTypes.mkString(", ")}, got ${actualDataTypes.mkString(", ")}")
+      batch.rowIterator.asScala
```

Comment:
Some explanation here:
The master impl has an issue,
the member "private var currentIter = ..." will be computed when create the returned iterator. Computing the `currentIter` require to read the first element of `columnarBatchIter`. But note that this block code being called from `EvalPythonExec.doExecute`, we should not read iterator here, we can only generate iterator. Reading iterator should start after the total iterator pipeline constructed.

Reply:
If I understand this correctly, you are saying that when the `Iterator` is constructed, the first batch of rows is read and that is wrong? I think you are right about that

Reply:
This is a separate issue that could be merged separately. I created https://issues.apache.org/jira/browse/SPARK-27968 and I will submit a PR to fix it.